# Importing Libraries and Files

In [3]:
import pandas as pd
import numpy as np
import os

In [5]:
# Creating path
path = r'C:\Users\shaiv\Downloads\Victoria Airbnb Analysis'

In [7]:
vic_listings = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'listings.csv'))

# Looking into the data set

In [10]:
# Setting number of columns 
pd.set_option('display.max_columns',None)

In [12]:
# Displaying 5 rows
vic_listings.head()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,host_url,host_name,host_since,host_location,host_about,host_response_time,host_response_rate,host_acceptance_rate,host_is_superhost,host_thumbnail_url,host_picture_url,host_neighbourhood,host_listings_count,host_total_listings_count,host_verifications,host_has_profile_pic,host_identity_verified,neighbourhood,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,amenities,price,minimum_nights,maximum_nights,minimum_minimum_nights,maximum_minimum_nights,minimum_maximum_nights,maximum_maximum_nights,minimum_nights_avg_ntm,maximum_nights_avg_ntm,calendar_updated,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,first_review,last_review,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,19898,https://www.airbnb.com/rooms/19898,20241229042348,2024-12-29,previous scrape,Duchess Suite,NaN,NaN,https://a0.muscache.com/pictures/101822/ae950f...,69729,https://www.airbnb.com/users/show/69729,Ryan,2010-01-11,NaN,NaN,NaN,NaN,NaN,f,https://a0.muscache.com/im/users/69729/profile...,https://a0.muscache.com/im/users/69729/profile...,NaN,1,6,"['email', 'phone']",t,f,NaN,Rockland,Victoria,48.42515,-123.33977,Private room in rental unit,Private room,1,NaN,1 bath,1.0,NaN,"[""Free parking on premises"", ""Breakfast"", ""Wif...",NaN,2,30,2,2,30,30,2.0,30.0,NaN,NaN,0,0,0,0,2024-12-29,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,f,1,0,1,0,NaN
1,41168,https://www.airbnb.com/rooms/41168,20241229042348,2024-12-29,city scrape,"Burma Suite, Ocean Island Garden Suites",Make Ocean Island Suites your home away from h...,"Our Suites are located in beautiful James Bay,...",https://a0.muscache.com/pictures/miso/Hosting-...,173169,https://www.airbnb.com/users/show/173169,Ocean Island Garden Suites,2010-07-19,"Victoria, Canada",“A seriously good deal for groups of up to fou...,within an hour,100%,99%,t,https://a0.muscache.com/im/pictures/user/682c0...,https://a0.muscache.com/im/pictures/user/682c0...,James Bay,3,6,"['email', 'phone']",t,f,Neighborhood highlights,James Bay,Victoria,48.41333,-123.37065,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,1.0,"[""Free parking on premises"", ""Outdoor furnitur...","$9,999.00",1,90,1,1,90,90,1.0,90.0,NaN,t,0,0,0,0,2024-12-29,70,9,0,2013-07-04,2024-09-29,4.29,4.48,4.42,4.52,4.57,4.83,4.23,38790,t,3,3,0,0,0.50
2,83515,https://www.airbnb.com/rooms/83515,20241229042348,2024-12-29,city scrape,GALLERY B&B SUITE,A quiet and sunny suite with a private entranc...,"Our property has gardens, an orchard, a pond a...",https://a0.muscache.com/pictures/6110737/930db...,455118,https://www.airbnb.com/users/show/455118,Celia,2011-03-21,"British Columbia, Canada",Your hosts Nicholas Hunt and Celia Duthie have...,within an hour,100%,97%,t,https://a0.muscache.com/im/users/455118/profil...,https://a0.muscache.com/im/users/455118/profil...,NaN,2,4,"['email', 'phone']",t,t,Neighborhood highlights,Salt Spring Island,Salt Spring Island,48.86222,-123.49996,Entire guest suite,Entire home/apt,2,1.0,1 bath,1.0,0.0,"[""Free parking on premises"", ""Dishes and silve...",$175.00,2,730,2,2,730,730,2.0,730.0,NaN,t,8,23,44,289,2024-12-29,148,13,0,2011-05-05,2024-10-13,4.88,4.89,4.87,4.92,4.84,4.94,4.75,NaN,t,2,2,0,0,0.89
3,168460,https://www.airbnb.com/rooms/168460,20241229042348,2024-12-29,city scrape,Perfectly located lovely home,"Wonderful neighborhood, two blocks to inner ha...",Lovely neighborhood with friendly neighbors su...,

In [14]:
# Checking the shape
vic_listings.shape

(3951, 75)

In [16]:
# List of columns
vic_listings.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

##### Many columns are not needed for further analysis.

In [19]:
# Checking data types
pd.set_option('display.max_rows', None)
vic_listings.dtypes

id                                                int64
listing_url                                      object
scrape_id                                         int64
last_scraped                                     object
source                                           object
name                                             object
description                                      object
neighborhood_overview                            object
picture_url                                      object
host_id                                           int64
host_url                                         object
host_name                                        object
host_since                                       object
host_location                                    object
host_about                                       object
host_response_time                               object
host_response_rate                               object
host_acceptance_rate                            

In [21]:
# Describing the data
vic_listings.describe()

,id,scrape_id,host_id,host_listings_count,host_total_listings_count,latitude,longitude,accommodates,bathrooms,bedrooms,beds,minimum_nights,maximum_nights,minimum_minimum_nights,maximum_minimum_nights,minimum_maximum_nights,maximum_maximum_nights,minimum_nights_avg_ntm,maximum_nights_avg_ntm,calendar_updated,availability_30,availability_60,availability_90,availability_365,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
count,3.951000e+03,3.951000e+03,3.951000e+03,3951.000000,3951.000000,3951.000000,3951.000000,3951.000000,3300.000000,3863.000000,3299.000000,3951.000000,3951.000000,3951.000000,3951.000000,3951.000000,3.951000e+03,3951.000000,3.951000e+03,0.0,3951.000000,3951.000000,3951.000000,3951.000000,3951.000000,3951.000000,3951.000000,3474.000000,3474.00000,3474.000000,3474.000000,3474.000000,3474.000000,3474.000000,3951.000000,3951.000000,3951.000000,3951.000000,3474.000000
mean,4.928531e+17,2.024123e+13,1.703816e+08,17.973931,25.926095,48.513728,-123.486788,3.806631,1.314091,1.614031,2.105487,14.253354,420.820046,14.140471,16.421919,611.612503,5.441570e+05,15.799595,5.429173e+05,NaN,13.768666,29.580866,46.263731,180.920273,64.532270,14.231081,0.537585,4.845826,4.85979,4.843520,4.898630,4.899594,4.864750,4.762052,5.059226,4.208302,0.842572,0.007846,1.793572
std,5.001071e+17,2.199497e+00,1.804464e+08,116.983340,153.632200,0.150023,0.230249,2.235743,0.675132,1.051437,1.438373,28.794037,433.582097,29.078658,30.772426,475.083224,3.416459e+07,30.255497,3.408730e+07,NaN,12.340451,24.610803,36.261271,127.956316,101.582501,19.613997,1.233926,0.262663,0.25823,0.272431,0.213698,0.240493,0.209329,0.291724,9.822941,9.558714,3.360533,0.088241,1.834674
min,1.989800e+04,2.024123e+13,1.329300e+04,1.000000,1.000000,48.328820,-124.424091,1.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,2.000000e+00,1.000000,2.000000e+00,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.00000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.010000
25%,3.318418e+07,2.024123e+13,2.679771e+07,1.000000,1.000000,48.423332,-123.517505,2.000000,1.000000,1.000000,1.000000,1.000000,30.000000,1.000000,2.000000,90.000000,1.000000e+02,2.000000,1.000000e+02,NaN,0.000000,0.000000,1.000000,61.000000,5.000000,0.000000,0.000000,4.810000,4.83000,4.810000,4.890000,4.890000,4.830000,4.700000,1.000000,1.000000,0.000000,0.000000,0.440000
50%,5.911464e+17,2.024123e+13,9.326026e+07,1.000000,2.000000,48.450400,-123.408075,4.000000,1.000000,1.000000,2.000000,2.000000,365.000000,2.000000,2.000000,365.000000,3.650000e+02,2.000000,3.650000e+02,NaN,15.000000,32.000000,54.000000,179.000000,26.000000,6.000000,0.000000,4.920000,4.93000,4.930000,4.960000,4.960000,4.910000,4.830000,1.000000,1.000000,0.000000,0.000000,1.210000
75%,9.533839e+17,2.024123e+13,2.753436e+08,4.000000,6.000000,48.553488,-123.362500,5.000000,1.500000,2.000000,3.000000,7.000000,999.000000,7.000000,28.000000,1125.000000,1.125000e+03,14.000000,1.125000e+03,NaN,27.000000,55.000000,83.000000,301.000000,80.000000,21.000000,1.000000,5.000000,4.99000,5.000000,5.000000,5.000000,4.980000,4.920000,3.000000,2.000000,0.000000,0.000000,2.550000
max,1.318260e+18,2.024123e+13,6.680625e+08,1419.000000,1845.000000,48.996710,-123.066340,16.000000,7.000000,14.000000,12.000000,365.000000,1125.000000,365.000000,365.000000,1125.000000,2.147484e+09,365.000000,2.142625e+09,NaN,30.000000,60.000000,90.000000,365.000000,909.000000,161.000000,12.000000,5.000000,5.00000,5.000000,5.000000,5.000000,5.000000,5.000000,53.000000,53.000000,29.000000,1.000000,14.490000


# Data Wrangling

In [24]:
# Removing columns that won't be useful for further analysis of the project
vic_listings = vic_listings.drop(columns = ['listing_url','picture_url', 'neighborhood_overview','last_scraped', 'source',
                                                    'description','name','host_url', 'host_name', 'host_about','host_thumbnail_url',
                                                    'host_location','host_picture_url', 'first_review', 'last_review',
                                                    'host_neighbourhood', 'host_listings_count', 'host_total_listings_count',
                                                    'host_verifications', 'host_has_profile_pic', 'minimum_minimum_nights',
                                                    'maximum_minimum_nights', 'minimum_maximum_nights','maximum_maximum_nights', 
                                                    'minimum_nights_avg_ntm','maximum_nights_avg_ntm', 'license', 'calendar_updated',
                                                   ])

In [26]:
# Checking the shape
vic_listings.shape

(3951, 47)

In [28]:
vic_listings.head()

,id,scrape_id,host_id,host_since,host_response_time,host_response_rate,host_acceptance_rate,host_is_superhost,host_identity_verified,neighbourhood,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,amenities,price,minimum_nights,maximum_nights,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,19898,20241229042348,69729,2010-01-11,NaN,NaN,NaN,f,f,NaN,Rockland,Victoria,48.42515,-123.33977,Private room in rental unit,Private room,1,NaN,1 bath,1.0,NaN,"[""Free parking on premises"", ""Breakfast"", ""Wif...",NaN,2,30,NaN,0,0,0,0,2024-12-29,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,f,1,0,1,0,NaN
1,41168,20241229042348,173169,2010-07-19,within an hour,100%,99%,t,f,Neighborhood highlights,James Bay,Victoria,48.41333,-123.37065,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,1.0,"[""Free parking on premises"", ""Outdoor furnitur...","$9,999.00",1,90,t,0,0,0,0,2024-12-29,70,9,0,4.29,4.48,4.42,4.52,4.57,4.83,4.23,t,3,3,0,0,0.50
2,83515,20241229042348,455118,2011-03-21,within an hour,100%,97%,t,t,Neighborhood highlights,Salt Spring Island,Salt Spring Island,48.86222,-123.49996,Entire guest suite,Entire home/apt,2,1.0,1 bath,1.0,0.0,"[""Free parking on premises"", ""Dishes and silve...",$175.00,2,730,t,8,23,44,289,2024-12-29,148,13,0,4.88,4.89,4.87,4.92,4.84,4.94,4.75,t,2,2,0,0,0.89
3,168460,20241229042348,802981,2011-07-10,within a few hours,100%,NaN,f,f,Neighborhood highlights,James Bay,Victoria,48.41620,-123.36790,Entire townhouse,Entire home/apt,3,2.5,2.5 baths,2.0,2.0,"[""Free parking on premises"", ""Dishes and silve...",$200.00,60,180,t,23,53,83,173,2024-12-29,64,0,0,4.91,4.90,4.98,4.89,4.92,5.00,4.90,f,1,1,0,0,0.42
4,185732,20241229042348,880366,2011-07-27,a few days or more,10%,0%,f,t,NaN,Highlands,Highlands,48.50589,-123.46357,Private room in home,Private room,5,1.0,1 private bath,2.0,2.0,"[""Free parking on premises"", ""Hair dryer"", ""Sm...",$199.00,1,365,t,30,60,90,365,2024-12-29,52,0,0,4.72,4.67,4.65,4.82,4.88,4.61,4.61,f,3,0,3,0,0.32


In [30]:
# Renaming 4 columns
vic_listings = vic_listings.rename(columns = {'calculated_host_listings_count' : 'listings_count',
                                         'calculated_host_listings_count_entire_homes': 'listings_count_home',
                                         'calculated_host_listings_count_private_rooms': 'listings_count_p_rooms',
                                         'calculated_host_listings_count_shared_rooms': 'listings_count_s_rooms'})

# Data Consistency Checks

### Handling Missing Values

In [34]:
# Checking for missing values
vic_listings.isnull().sum()

id                                 0
scrape_id                          0
host_id                            0
host_since                         0
host_response_time               844
host_response_rate               844
host_acceptance_rate             398
host_is_superhost                111
host_identity_verified             0
neighbourhood                   1630
neighbourhood_cleansed             0
neighbourhood_group_cleansed       0
latitude                           0
longitude                          0
property_type                      0
room_type                          0
accommodates                       0
bathrooms                        651
bathrooms_text                     1
bedrooms                          88
beds                             652
amenities                          0
price                            650
minimum_nights                     0
maximum_nights                     0
has_availability                  52
availability_30                    0
a

##### Note: The neighbourhood column has 41.25% missing data, so it will be dropped as the majority of its values are missing, making it less useful for analysis. The price column has 16.45% missing data, which can be imputed using the mean based on relevant features. However, since the objective is to predict price using correlated features, removing rows with missing price values is a better approach to avoid introducing bias through imputation. It is important to note that this decision may lead to bias in the dataset, potentially impacting the accuracy of the results. Columns with less than 5% missing data will be retained since the amount of missing data is insignificant. For numerical columns with more than 5% missing data, imputation will be considered if it is necessary for the analysis. The columns bathrooms_text and bathrooms mean the same thing, so we will remove bathrooms_text. The host_response_time is not required for further analysis.

In [37]:
# Removing neighborhood column
vic_listings = vic_listings.drop(columns = ['neighbourhood', 'host_response_time', 'bathrooms_text'])

In [39]:
# Removing missing prices row
vic_listings = vic_listings[~vic_listings['price'].isnull()]

In [41]:
# Removing $ and comma(,) from the price column so we can extract the actual number when we want to perform aggregation when required
vic_listings['price'] = vic_listings['price'].replace({'\$' : '' , ',' : ''}, regex = True)

<>:2: SyntaxWarning: invalid escape sequence '\$'
<>:2: SyntaxWarning: invalid escape sequence '\$'
C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\2868051269.py:2: SyntaxWarning: invalid escape sequence '\$'
  vic_listings['price'] = vic_listings['price'].replace({'\$' : '' , ',' : ''}, regex = True)


In [43]:
vic_listings['price'].value_counts()

price
100.00     93
150.00     70
120.00     69
200.00     66
110.00     60
250.00     51
90.00      50
175.00     50
125.00     45
99.00      45
95.00      42
130.00     40
105.00     34
85.00      34
115.00     33
140.00     32
135.00     31
89.00      31
300.00     29
160.00     29
126.00     26
225.00     24
350.00     24
145.00     24
180.00     24
75.00      22
98.00      22
108.00     21
80.00      21
136.00     21
129.00     21
50.00      21
111.00     21
70.00      20
190.00     20
400.00     20
189.00     19
149.00     19
165.00     19
199.00     19
76.00      19
106.00     19
128.00     18
170.00     18
185.00     18
121.00     18
159.00     17
155.00     17
109.00     17
96.00      17
65.00      17
113.00     16
169.00     16
114.00     16
88.00      15
500.00     15
86.00      15
220.00     15
101.00     15
302.00     15
195.00     15
60.00      14
103.00     14
179.00     14
153.00     14
79.00      14
131.00     14
104.00     13
134.00     12
55.00      12
450.00     12


In [45]:
# Converting price to float
vic_listings['price'] = vic_listings['price'].astype('float')

In [47]:
# Converting listing_count to float
vic_listings['listings_count'] = vic_listings['listings_count'].astype('float')

In [49]:
# Removing % sign and convert acceptance rate to float
vic_listings['host_acceptance_rate'] = vic_listings['host_acceptance_rate'].replace({'\%':''},regex = True).astype('float')

<>:2: SyntaxWarning: invalid escape sequence '\%'
<>:2: SyntaxWarning: invalid escape sequence '\%'
C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\918540606.py:2: SyntaxWarning: invalid escape sequence '\%'
  vic_listings['host_acceptance_rate'] = vic_listings['host_acceptance_rate'].replace({'\%':''},regex = True).astype('float')


In [51]:
# Adding mean values to host_acceptance_rate
vic_listings.loc[vic_listings['host_acceptance_rate'].isnull(),'host_acceptance_rate'] = vic_listings['host_acceptance_rate'].mean()

In [53]:
# Removing % sign and convert response rate to float
vic_listings['host_response_rate'] = vic_listings['host_response_rate'].replace({'\%':''},regex = True).astype('float')

<>:2: SyntaxWarning: invalid escape sequence '\%'
<>:2: SyntaxWarning: invalid escape sequence '\%'
C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\2770780378.py:2: SyntaxWarning: invalid escape sequence '\%'
  vic_listings['host_response_rate'] = vic_listings['host_response_rate'].replace({'\%':''},regex = True).astype('float')


In [55]:
# Adding mean values to host_response_rate
vic_listings.loc[vic_listings['host_response_rate'].isnull(),'host_response_rate'] = vic_listings['host_response_rate'].mean()

In [57]:
# Adding mean values to review_scores_value
vic_listings.loc[vic_listings['review_scores_value'].isnull(),'review_scores_value'] = vic_listings['review_scores_value'].mean()

In [59]:
# Adding mean values to review_scores_rating
vic_listings.loc[vic_listings['review_scores_rating'].isnull(),'review_scores_rating'] = vic_listings['review_scores_rating'].mean()

In [61]:
# Adding mean values to review_scores_accuracy
vic_listings.loc[vic_listings['review_scores_accuracy'].isnull(),'review_scores_accuracy'] = vic_listings['review_scores_accuracy'].mean()

In [63]:
# Adding mean values to review_scores_cleanliness
vic_listings.loc[vic_listings['review_scores_cleanliness'].isnull(),'review_scores_cleanliness'] = vic_listings['review_scores_cleanliness'].mean()

In [65]:
# Adding mean values to review_scores_checkin
vic_listings.loc[vic_listings['review_scores_checkin'].isnull(),'review_scores_checkin'] = vic_listings['review_scores_checkin'].mean()

In [67]:
# Adding mean values to review_scores_communication
vic_listings.loc[vic_listings['review_scores_communication'].isnull(),'review_scores_communication'] = vic_listings['review_scores_communication'].mean()

In [69]:
# Adding mean values to review_scores_location
vic_listings.loc[vic_listings['review_scores_location'].isnull(),'review_scores_location'] = vic_listings['review_scores_location'].mean()

In [71]:
# Adding mean values to reviews_per_month
vic_listings.loc[vic_listings['reviews_per_month'].isnull(),'reviews_per_month'] = vic_listings['reviews_per_month'].mean()

In [73]:
# Count of bathrooms
vic_listings['bathrooms'].value_counts(dropna = False)

bathrooms
1.0    2427
2.0     463
1.5     143
2.5      94
3.0      76
3.5      27
4.0      19
0.5      16
0.0      11
4.5       8
6.0       5
5.0       5
5.5       3
6.5       2
NaN       1
7.0       1
Name: count, dtype: int64

In [75]:
# The frequency distribution shows that maximum listings have 1 bathroom, so we will fill null value with 1
vic_listings['bathrooms'] = vic_listings['bathrooms'].fillna(1)

In [77]:
# Count of bedrooms
vic_listings['bedrooms'].value_counts(dropna = False)

bedrooms
1.0     1724
2.0      886
3.0      310
0.0      216
4.0      100
5.0       41
6.0       19
7.0        4
14.0       1
Name: count, dtype: int64

In [79]:
# The frequency distribution shows that maximum listings have 1 bedrooms, so will fill null value with 1 and an outlier '14' will be replace with 1
vic_listings['bedrooms'] = vic_listings['bedrooms'].fillna(1)
vic_listings.loc[vic_listings['bedrooms'] == 14, 'bedrooms'] = 1

In [81]:
# Count of beds
vic_listings['beds'].value_counts(dropna = False)

beds
1.0     1312
2.0     1041
3.0      437
4.0      251
5.0       95
0.0       56
6.0       54
8.0       18
7.0       17
9.0       10
10.0       5
12.0       2
NaN        2
11.0       1
Name: count, dtype: int64

In [83]:
# The frequency distribution shows that maximum listings have 1 beds, so we will fill null value with 1
vic_listings['beds'] = vic_listings['beds'].fillna(1)

In [85]:
# Count of host_is_superhost
vic_listings['host_is_superhost'].value_counts(dropna = False)

host_is_superhost
t      1807
f      1405
NaN      89
Name: count, dtype: int64

In [87]:
# The frequency distribution shows that maximum listings have 't', so we will fill null value with 't'
vic_listings.loc[vic_listings['host_is_superhost'].isnull(),'host_is_superhost'] = 't'

In [89]:
# Changing boolean value from t and f to 1 and 0
vic_listings.loc[vic_listings['host_is_superhost'] == 't', 'host_is_superhost'] = 1
vic_listings.loc[vic_listings['host_is_superhost'] == 'f', 'host_is_superhost'] = 0

In [91]:
# Count of has_availability
vic_listings['has_availability'].value_counts(dropna = False)

has_availability
t      3284
NaN      17
Name: count, dtype: int64

In [93]:
# The frequency distribution shows that maximum listings have 't', so we will fill null value with 't'. None of the availability are 'f', so it is possible that all locations might be available.
vic_listings.loc[vic_listings['has_availability'].isnull(),'has_availability'] = 't'

In [95]:
# Changing boolean value from t to 1  (There are no f values in this column)
vic_listings.loc[vic_listings['has_availability'] == 't', 'has_availability'] = 1

In [97]:
# Changing boolean value from t and f to 1 and 0 (There are no f values in this column)
vic_listings.loc[vic_listings['host_is_superhost'] == 't', 'host_is_superhost'] = 1
vic_listings.loc[vic_listings['host_is_superhost'] == 'f', 'host_is_superhost'] = 0

In [99]:
# Changing boolean value from t and f to 1 and 0
vic_listings.loc[vic_listings['host_identity_verified'] == 't', 'host_identity_verified'] = 1
vic_listings.loc[vic_listings['host_identity_verified'] == 'f', 'host_identity_verified'] = 0

In [101]:
# Changing boolean value from t and f to 1 and 0
vic_listings.loc[vic_listings['instant_bookable'] == 't', 'instant_bookable'] = 1
vic_listings.loc[vic_listings['instant_bookable'] == 'f', 'instant_bookable'] = 0

# Checking for mixed type in orders

In [104]:
for col in vic_listings.columns.tolist():
  weird = (vic_listings[[col]].applymap(type) != vic_listings[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (vic_listings[weird]) > 0:
    print (col)

C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\4047048329.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (vic_listings[[col]].applymap(type) != vic_listings[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\4047048329.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (vic_listings[[col]].applymap(type) != vic_listings[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\4047048329.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (vic_listings[[col]].applymap(type) != vic_listings[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Temp\ipykernel_18812\4047048329.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (vic_listings[[col]].applymap(type) != vic_listings[[col]].iloc[0].apply(type)).an

##### There are no mixed datatypes in the dataset

In [107]:
# Resetting number of rows to display
pd.reset_option('max_rows')

In [109]:
# Checking for missing values
vic_listings.isnull().sum()

id                              0
scrape_id                       0
host_id                         0
host_since                      0
host_response_rate              0
host_acceptance_rate            0
host_is_superhost               0
host_identity_verified          0
neighbourhood_cleansed          0
neighbourhood_group_cleansed    0
latitude                        0
longitude                       0
property_type                   0
room_type                       0
accommodates                    0
bathrooms                       0
bedrooms                        0
beds                            0
amenities                       0
price                           0
minimum_nights                  0
maximum_nights                  0
has_availability                0
availability_30                 0
availability_60                 0
availability_90                 0
availability_365                0
calendar_last_scraped           0
number_of_reviews               0
number_of_revi

# Check for duplicate data

In [112]:
vic_listings.duplicated().sum()

0

##### There are no duplicate entries in the dataset.

# Understanding the data

In [116]:
# Descriptive statistics for data
vic_listings.describe()

,id,scrape_id,host_id,host_response_rate,host_acceptance_rate,latitude,longitude,accommodates,bathrooms,bedrooms,beds,price,minimum_nights,maximum_nights,availability_30,availability_60,availability_90,availability_365,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,listings_count,listings_count_home,listings_count_p_rooms,listings_count_s_rooms,reviews_per_month
count,3.301000e+03,3.301000e+03,3.301000e+03,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000,3301.000000
mean,5.164819e+17,2.024123e+13,1.751980e+08,96.880282,87.101145,48.515786,-123.494869,3.812784,1.313996,1.567404,2.104817,213.322630,13.981521,422.813087,16.055135,34.558013,54.054832,211.661315,66.688882,15.617692,0.620721,4.847149,4.859450,4.845853,4.899157,4.901152,4.865502,4.760873,5.248410,4.356559,0.882763,0.008482,1.894828
std,5.071577e+17,1.844029e+00,1.838539e+08,13.685308,23.738052,0.151296,0.237684,2.261154,0.675052,1.023105,1.438195,353.792884,27.126582,430.345793,11.882607,23.204925,33.498345,113.572224,103.925541,20.375903,1.311158,0.247927,0.246807,0.258407,0.198458,0.215935,0.195967,0.272946,9.897724,9.610873,3.514629,0.091722,1.760168
min,4.116800e+04,2.024123e+13,1.329300e+04,0.000000,0.000000,48.328820,-124.424091,1.000000,0.000000,0.000000,0.000000,29.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.010000
25%,3.385434e+07,2.024123e+13,2.858359e+07,100.000000,87.101145,48.423330,-123.521264,2.000000,1.000000,1.000000,1.000000,106.000000,1.000000,31.000000,1.000000,8.000000,23.000000,114.000000,5.000000,1.000000,0.000000,4.830000,4.850000,4.830000,4.899157,4.901152,4.840000,4.730000,1.000000,1.000000,0.000000,0.000000,0.600000
50%,6.170352e+17,2.024123e+13,9.488487e+07,100.000000,98.000000,48.451015,-123.412280,4.000000,1.000000,1.000000,2.000000,150.000000,2.000000,365.000000,19.000000,43.000000,64.000000,226.000000,27.000000,8.000000,0.000000,4.900000,4.910000,4.900000,4.940000,4.950000,4.890000,4.800000,1.000000,1.000000,0.000000,0.000000,1.610000
75%,9.900101e+17,2.024123e+13,2.972652e+08,100.000000,100.000000,48.554430,-123.364170,4.000000,1.500000,2.000000,3.000000,233.000000,7.000000,730.000000,27.000000,56.000000,85.000000,319.000000,82.000000,23.000000,1.000000,4.980000,4.980000,4.990000,5.000000,5.000000,4.970000,4.910000,3.000000,2.000000,0.000000,0.000000,2.450000
max,1.318260e+18,2.024123e+13,6.680625e+08,100.000000,100.000000,48.996710,-123.066340,16.000000,7.000000,7.000000,12.000000,9999.000000,365.000000,1125.000000,30.000000,60.000000,90.000000,365.000000,909.000000,161.000000,12.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,53.000000,53.000000,29.000000,1.000000,14.490000


##### The maximum price of 9999 is very high. It might be a big property or error in the pricing. The maximum value for minimum_nights and maximum_nights are 365 and 1125, which seems pretty high. We should check for outliers in the all these three columns.

In [119]:
# Count of price
pd.set_option('display.max_rows', None)
vic_listings['price'].value_counts(dropna = False).sort_index()

price
29.0       1
30.0       1
32.0       2
33.0       1
35.0       1
38.0       2
39.0       1
40.0       6
41.0       1
42.0       2
44.0       1
45.0       4
46.0       1
48.0       4
50.0      21
51.0       3
52.0       3
53.0       2
54.0       4
55.0      12
56.0       1
57.0       2
58.0       5
59.0       9
60.0      14
61.0       5
62.0       4
63.0       3
64.0       5
65.0      17
66.0       5
67.0       3
68.0       3
69.0       8
70.0      20
71.0       6
72.0      11
73.0       5
74.0       3
75.0      22
76.0      19
77.0       5
78.0       7
79.0      14
80.0      21
81.0       7
82.0      10
83.0       6
84.0       9
85.0      34
86.0      15
87.0       6
88.0      15
89.0      31
90.0      50
91.0       9
92.0       7
93.0       7
94.0      10
95.0      42
96.0      17
97.0       8
98.0      22
99.0      45
100.0     93
101.0     15
102.0      9
103.0     14
104.0     13
105.0     34
106.0     19
107.0      8
108.0     21
109.0     17
110.0     60
111.0     21
112.0 

In [121]:
# Check the listing for highest price property
vic_listings.loc[vic_listings['price'].isin([2000, 5806, 9956, 9999])]

,id,scrape_id,host_id,host_since,host_response_rate,host_acceptance_rate,host_is_superhost,host_identity_verified,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bedrooms,beds,amenities,price,minimum_nights,maximum_nights,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,listings_count,listings_count_home,listings_count_p_rooms,listings_count_s_rooms,reviews_per_month
1,41168,20241229042348,173169,2010-07-19,100.000000,99.000000,1,0,James Bay,Victoria,48.413330,-123.37065,Entire rental unit,Entire home/apt,2,1.0,1.0,1.0,"[""Free parking on premises"", ""Outdoor furnitur...",9999.0,1,90,1,0,0,0,0,2024-12-29,70,9,0,4.290000,4.48000,4.420000,4.520000,4.570000,4.830000,4.230000,1,3.0,3,0,0,0.500000
289,11243817,20241229042348,27217133,2015-02-04,96.880282,100.000000,1,1,Saanich,Saanich,48.483410,-123.31651,Entire home,Entire home/apt,2,1.0,0.0,1.0,"[""Free parking on premises"", ""Outdoor furnitur...",9956.0,2,120,1,23,52,81,260,2024-12-29,57,0,0,4.840000,4.96000,4.980000,4.950000,4.960000,4.820000,4.840000,0,2.0,2,0,0,0.530000
444,16599081,20241229042348,109495815,2017-01-02,0.000000,0.000000,0,1,Juan de Fuca,Juan de Fuca,48.556250,-124.41888,Entire cabin,Entire home/apt,12,2.5,4.0,12.0,"[""Free parking on premises"", ""Dishes and silve...",5806.0,2,1125,1,23,46,76,319,2024-12-29,63,0,0,4.860000,4.81000,4.630000,4.900000,4.900000,4.980000,4.630000,0,1.0,1,0,0,0.660000
727,24541465,20241229042348,173169,2010-07-19,100.000000,99.000000,1,0,James Bay,Victoria,48.413160,-123.37065,Entire rental unit,Entire home/apt,3,1.0,1.0,2.0,"[""Free parking on premises"", ""Outdoor furnitur...",9999.0,1,1125,1,0,0,0,0,2024-12-29,40,8,0,4.480000,4.70000,4.650000,4.580000,4.450000,4.830000,4.180000,0,3.0,3,0,0,0.600000
3913,1302997863073985323,20241229042348,658337247,2024-10-21,96.880282,87.101145,0,1,Langford,Langford,48.457147,-123.53654,Entire rental unit,Entire home/apt,4,2.0,2.0,2.0,"[""TV"", ""Exercise equipment"", ""Kitchen"", ""Free ...",2000.0,1,365,1,30,60,90,365,2024-12-29,0,0,0,4.847149,4.85945,4.845853,4.899157,4.901152,4.865502,4.760873,0,1.0,1,0,0,1.894828


In [123]:
# Checking for similar property for the highest property prices
vic_listings.loc[vic_listings['id'].isin([41168, 11243817, 16599081, 24541465, 1302997863073985323])]

,id,scrape_id,host_id,host_since,host_response_rate,host_acceptance_rate,host_is_superhost,host_identity_verified,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bedrooms,beds,amenities,price,minimum_nights,maximum_nights,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,listings_count,listings_count_home,listings_count_p_rooms,listings_count_s_rooms,reviews_per_month
1,41168,20241229042348,173169,2010-07-19,100.000000,99.000000,1,0,James Bay,Victoria,48.413330,-123.37065,Entire rental unit,Entire home/apt,2,1.0,1.0,1.0,"[""Free parking on premises"", ""Outdoor furnitur...",9999.0,1,90,1,0,0,0,0,2024-12-29,70,9,0,4.290000,4.48000,4.420000,4.520000,4.570000,4.830000,4.230000,1,3.0,3,0,0,0.500000
289,11243817,20241229042348,27217133,2015-02-04,96.880282,100.000000,1,1,Saanich,Saanich,48.483410,-123.31651,Entire home,Entire home/apt,2,1.0,0.0,1.0,"[""Free parking on premises"", ""Outdoor furnitur...",9956.0,2,120,1,23,52,81,260,2024-12-29,57,0,0,4.840000,4.96000,4.980000,4.950000,4.960000,4.820000,4.840000,0,2.0,2,0,0,0.530000
444,16599081,20241229042348,109495815,2017-01-02,0.000000,0.000000,0,1,Juan de Fuca,Juan de Fuca,48.556250,-124.41888,Entire cabin,Entire home/apt,12,2.5,4.0,12.0,"[""Free parking on premises"", ""Dishes and silve...",5806.0,2,1125,1,23,46,76,319,2024-12-29,63,0,0,4.860000,4.81000,4.630000,4.900000,4.900000,4.980000,4.630000,0,1.0,1,0,0,0.660000
727,24541465,20241229042348,173169,2010-07-19,100.000000,99.000000,1,0,James Bay,Victoria,48.413160,-123.37065,Entire rental unit,Entire home/apt,3,1.0,1.0,2.0,"[""Free parking on premises"", ""Outdoor furnitur...",9999.0,1,1125,1,0,0,0,0,2024-12-29,40,8,0,4.480000,4.70000,4.650000,4.580000,4.450000,4.830000,4.180000,0,3.0,3,0,0,0.600000
3913,1302997863073985323,20241229042348,658337247,2024-10-21,96.880282,87.101145,0,1,Langford,Langford,48.457147,-123.53654,Entire rental unit,Entire home/apt,4,2.0,2.0,2.0,"[""TV"", ""Exercise equipment"", ""Kitchen"", ""Free ...",2000.0,1,365,1,30,60,90,365,2024-12-29,0,0,0,4.847149,4.85945,4.845853,4.899157,4.901152,4.865502,4.760873,0,1.0,1,0,0,1.894828


##### We will remove the rows where the price is 9956 and 9999 as they are clear outliers, and the hosts do not have any other listings to verify the pricing. Additionally, the prices appear excessively high compared to typical property sizes. However, the properties with prices of 2000 and 5806 seem reasonable based on the property size and could be justified by the neighbourhood's value, so these rows will be retained.

In [126]:
# Removing the rows for price 9956 and 9999
vic_listings = vic_listings.loc[~vic_listings['price'].isin([9956, 9999])]

In [128]:
# Count of minimum_nights
pd.set_option('display.max_rows', None)
vic_listings['minimum_nights'].value_counts(dropna = False).sort_index()

minimum_nights
1       869
2      1110
3       346
4        62
5        58
6        15
7        64
8         1
10       11
12        1
13        1
14       13
15        1
20        3
21        2
27        3
28       30
29       21
30      376
31       18
60       10
80        1
89       17
90      204
91       47
100       1
120       4
130       1
150       1
179       1
180       4
182       1
365       1
Name: count, dtype: int64

In [130]:
# Check the listing for minimum_nights at 365
vic_listings.loc[vic_listings['minimum_nights'].isin([365])]

,id,scrape_id,host_id,host_since,host_response_rate,host_acceptance_rate,host_is_superhost,host_identity_verified,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bedrooms,beds,amenities,price,minimum_nights,maximum_nights,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,listings_count,listings_count_home,listings_count_p_rooms,listings_count_s_rooms,reviews_per_month
1243,40560883,20241229042348,51776328,2015-12-19,96.880282,87.101145,0,1,North Park,Victoria,48.42772,-123.35989,Entire condo,Entire home/apt,2,1.0,1.0,1.0,"[""Long term stays allowed"", ""Dishes and silver...",100.0,365,700,1,30,60,90,365,2024-12-29,0,0,0,4.847149,4.85945,4.845853,4.899157,4.901152,4.865502,4.760873,0,1.0,1,0,0,1.894828


In [132]:
# Count of maximum_nights
pd.set_option('display.max_rows', None)
vic_listings['maximum_nights'].value_counts(dropna = False).sort_index()

maximum_nights
2         4
3         5
4         7
5        22
6         6
7        97
8         4
9         2
10       25
11        2
12        6
13        2
14      133
15       17
16        2
20        8
21       19
25        6
26        6
27       25
28      119
29       39
30      245
31       50
32        3
33        3
35        9
37        1
40        9
42        4
45       19
47        1
50        5
56        2
59        1
60      103
62        6
63        1
65        5
70        1
75        3
80        1
89        8
90      164
91        3
92        8
93        4
95       11
100      18
111       1
115       1
118       1
120      79
122       1
125       5
140       1
150      14
155       2
158       1
160       5
170       1
179       3
180      85
181       1
182       5
185       5
186       1
190       8
200      26
210       3
230       1
240       2
242       1
250       4
270       2
275       1
280       1
300       6
307       1
320       1
321       1
335       3
3

##### We will retain the row with minimum_nights set to 365 since the maximum_nights booked for that listing ID is 700, which exceeds the minimum nights requirement. Additionally, the maximum_nights count indicates that several bookings extend beyond 365 nights, suggesting long-term stays. There are no outliers in the maximum_nights column, as evidenced by the fact that 760 bookings were made for up to 1125 nights, indicating a consistent trend for extended stays.

In [135]:
# Checking the shape
vic_listings.shape

(3298, 44)

# Exporting Dataset

In [138]:
vic_listings.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'victoria_cleaned.csv'),index = False)